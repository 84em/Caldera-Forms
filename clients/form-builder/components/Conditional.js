import React, {Fragment} from 'react';

/**
 * The section to set which fields it applies to
 *
 * @since 1.8.10
 *
 * @param Array formFields The form's fields
 * @param Array fieldsUsed The fields this conditional is applied to.
 * @param Array notAllowedFields The fields this conditional can NOT applied to.
 * @param strings Translation strings
 * @param onChange
 * @param groupId
 * @returns {*}
 * @constructor
 */
export const AppliesToFields = ({formFields, fieldsUsed, notAllowedFields, strings, onChange, groupId}) => {
    /**
     * Check if we should disable this option
     *
     * @since 1.8.10
     *
     * @param fieldId
     * @returns {*}
     */
    function isFieldDisabled(fieldId) {
        if (undefined === notAllowedFields) {
            return false;
        }
        return notAllowedFields.includes(fieldId);
    }

    /**
     * Check if field is used and should be checked.
     *
     * @since 1.8.10
     *
     * @param fieldId
     * @returns {*}
     */
    function isFieldChecked(fieldId) {
        return fieldsUsed.includes(fieldId);
    }

    return (
        <div style={{float: 'left', width: '288px', paddingLeft: '12px'}}>
            <h4 style={{borderBottom: `1px solid rgb(191, 191, 191)`, margin: `0px 0px 6px; padding: 0px 0px 6px`}}>
                {strings['applied-fields']}
            </h4>
            <p className="description">{strings['select-apply-fields']}</p>
            {formFields.map(field => (
                    <label style={{display: 'block', marginLeft: `20px`}}>
                        {field.label}
                        <input
                            style={{marginLeft: `-20px`}}
                            type="checkbox"
                            disabled={isFieldDisabled(field.ID)}
                            onClick={(e) => {
                                e.preventDefault();
                                //Is it already applied?
                                if (isFieldChecked(field.ID)) {
                                    //remove from applied fields
                                    onChange(fieldsUsed.filter(f => f.ID === field.ID))
                                } else {
                                    //Add to applied fields.
                                    onChange([...fieldsUsed, field.ID])
                                }
                            }}
                            value={groupId}
                            checked={isFieldChecked(field.ID)}
                        />
                    </label>
                )
            )}
        </div>
    );
};

const isFieldTypeWithOptions = (fieldType) => ['dropdown', 'checkbox', 'radio', 'filtered_select2', 'toggle_switch'];


/**
 * One line in a group
 *
 * @since 1.8.10
 *
 * @param line The single line of conditional rule.
 * @param {} strings Translation field
 * @param boolean isFirst
 * @param formFields
 * @param onRemoveLine
 * @param onUpdateLine
 * @returns {*}
 * @constructor
 */
export const ConditionalLine = ({line, strings, isFirst, formFields, onRemoveLine, onUpdateLine}) => {
    const {value, field, compare} = line;
    const lineId = line.id;
    const fieldConfig = React.useMemo(() => {
        const config = formFields.find(f => f.ID === field);
        if (config) {
            return config;
        }
        return {
            type: 'text'
        }
    }, [field]);

    /**
     * Callback for changing which field this line is based on
     *
     * @param field
     */
    const onChangeField = (field) => {
        onUpdateLine({
            field, compare, value,
        })
    };

    /**
     * Callback for changing a line's value
     *
     * @since 1.8.10
     *
     * @param value
     */
    const onChangeValue = (value) => {
        onUpdateLine({
            field, compare, value,
        })
    };

    /**
     * Callback for changing a line's comparison type
     *
     * @since 1.8.10
     *
     * @param compare
     */
    const onChangeCompare = (compare) => {
        onUpdateLine({
            field, compare, value,
        })
    };


    return (
        <div key={lineId} className={`caldera-condition-line condition-line-${lineId}`}>
            <span style={{display: "inline-block"}}>
                {isFirst ? <React.Fragment>{strings['if']}</React.Fragment> :
                    <React.Fragment>{strings['and']}</React.Fragment>}
            </span>
            <select
                style={{maxWidth: "120px", verticalAlign: 'inherit'}}
                className="condition-line-field"
                value={field}
                onChange={(e) => onChangeField(e.target.value,lineId)}
            >
                <option/>
                <optgroup label={strings['fields']}>
                    {formFields.map(field => (
                        <option value={field.ID}>
                            {field.label} [{field.slug}]
                        </option>
                    ))}
                </optgroup>
            </select>
            <select
                className="condition-line-compare"
                value={compare}
                onChange={(e) => onChangeCompare(e.target.value)}
                style={{maxWidth: "120px", verticalAlign: 'inherit'}}
            >
                {['is', 'isnot', 'greater', 'smaller', 'startswith', 'endswith', 'contains'].map(compareType => (
                    <option key={compareType} value={compareType}>{compareType}</option>
                ))}
            </select>
            {isFieldTypeWithOptions(fieldConfig.type) ? (
                <select
                    value={value}
                    onChange={(e) => onChangeValue(e.target.value)}
                    style={{maxWidth: "165px", verticalAlign: 'inherit'}}>
                    <option/>
                    {formFields.map(option => {
                        //should be fieldConfig.option, except that doesn't exist.
                        return (
                            <option
                                key={option.value}
                                value={option.value}
                            >
                                {option.label}
                            </option>
                        )
                    })}
                </select>
            ) : (
                <input
                    onChange={onChangeValue}
                    value={value}
                    type="text"
                    className="magic-tag-enabled block-input"
                />
            )}
            <button
                onClick={(e) => {
                    e.preventDefault();
                    onRemoveLine(lineId);
                }}
                className="button pull-right" type="button"
            >
                <i className="icon-join"/>
            </button>
        </div>
    );
};

/**
 * All of the lines of one conditional
 *
 * @since 1.8.10
 *
 * @param lines
 * @param strings
 * @param formFields
 * @param onRemoveLine
 * @param onUpdateLine
 * @returns {*}
 * @constructor
 */
const ConditionalLines = ({lines, strings, formFields, onRemoveLine, onUpdateLine}) => {
    return (
        <div className="caldera-condition-group caldera-condition-lines">
            {lines.map((line, index) => {
                return (
                    <ConditionalLine
                        line={line}
                        isFirst={0 === index}
                        onUpdateLine={(update)=> onUpdateLine(update,line.id)}
                        {...{
                            strings,
                            formFields,
                            onRemoveLine,
                        }}
                    />
                )
            })}

            <div style={{margin: "12px 0 0"}}>
                <button
                    className="button button-small condition-group-add-line"
                    type="button"
                >
                    {strings['add-condition']}
                </button>
            </div>
        </div>)
};

/**
 * One conditional
 *
 * @since 1.8.10
 */
const Conditional = ({conditional, formFields, strings, onRemoveConditional, onUpdateConditional, fieldsNotAllowed, fieldsUsed}) => {
    const {type, config, id} = conditional;
    const group = config && config.hasOwnProperty('group') ? config.group : {};
    const name = conditional.hasOwnProperty('config') && conditional.config.hasOwnProperty('name') ? conditional.config.name : '';
    const onAddLine = (groupId) => {
        const id = `cl_${Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 5)}`
        onUpdateConditional({
            conditional,
            config: {
                ...conditional.config,
                group: {
                    ...group,
                    [groupId] : {
                        ...conditional.config[groupId],
                        [id]: {
                            field: '',
                            value: '',
                            type: ''
                        }
                    }

                }
            }
        });
    };

    /**
     * Callback for updating a line in a conditional group
     *
     * @since 1.8.10
     *
     * @param line
     * @param lineId
     */
    const onUpdateLine = (line, lineId,groupId) => {
        let group = conditional.config.group;
        onUpdateConditional({
            ...conditional,
            config: {
                ...conditional.config,
                group: {
                    ...group,
                    [groupId] :{
                        ...group[groupId],
                        [lineId]: line
                    }
                }
            }
        });
    };

    /**
     *  Callback for removing a line from a line group
     *  
     *  @since 1.8.10
     *
     * @param lineId
     * @param groupId
     */
    const onRemoveLine = (lineId, groupId) => {
        let groups = group;

        //Delete line from line group
        delete groups[groupId][lineId];

        //no lines left?
        if( ! Object.values(groups[groupId]).length){
            //delete line group
            delete groups[groupId];
        }
        onUpdateConditional({
            ...conditional,
            config: {
                ...conditional.config,
                group: groups
            }
        });
    };


    /**
     * Callback for updating name of conditional group
     *
     * @since 1.8.10
     *
     * @param name
     */
    const onUpdateName = (name) => {
        onUpdateConditional({
            ...conditional,
            config: {
                ...conditional.config,
                name,
            }
        })
    };

    /**
     * Callback for changing type of conditional
     *
     * @since 1.8.10
     *
     * @param type
     * @returns {*}
     */
    const onChangeType = (type) => {
        return onUpdateConditional({
            ...conditional,
            type
        })
    };


    return (
        <div className="caldera-editor-condition-config caldera-forms-condition-edit"
             style={{marginTop: '-27px', width: "auto"}}
        >

            <div
                className={`condition-point-${id}`}
                style={{float: 'left', width: "550px"}}
            >
                <div
                    className="caldera-config-group"
                >
                    <label htmlFor={`condition-group-name-${id}`}>
                        {strings.name}
                    </label>
                    <div
                        className="caldera-config-field"
                    >
                        <input
                            type="text"
                            name={`conditions[${id}][name]`}
                            id={`condition-group-name-${id}`}
                            value={name}
                            onChange={(e) => onUpdateName(e.target.value)}
                            required
                            className="required block-input condition-group-name"
                        />
                    </div>
                </div>

                <div className="caldera-config-group">
                    <label
                        htmlFor={`condition-group-type-${id}`}
                    >
                        {strings.type}
                    </label>
                    <div

                        className="caldera-config-field"
                    >
                        <select
                            value={type}
                            onChange={(e) => onChangeType(e.target.value)}
                            id={`condition-group-type-${id}`}
                            name={`conditions[${id}][type]`}
                            className="condition-group-type"
                        >
                            <option value=""/>
                            <option value="show">
                                {strings.show}
                            </option>
                            <option value="hide">
                                {strings.hide}
                            </option>
                            <option value="disable">
                                {strings.disable}
                            </option>
                        </select>
                        {type &&
                        <button
                            type="button"
                            className="pull-right button button-small condition-group-add-lines"
                            onClick={(e) => {
                                e.preventDefault();
                                onAddLine(groupId)
                            }}
                        >
                            {strings['add-conditional-line']}
                        </button>
                        }
                    </div>
                </div>
                {Object.keys(group).map(groupId => {
                    const lineGroups = group[groupId];
                    const lines = Object.keys(lineGroups).map(
                        (lineGroupId) => {
                            return {
                                ...lineGroups[lineGroupId],
                                id: lineGroupId,
                            }
                        }
                    );
                    return (
                        <Fragment>
                            <ConditionalLines
                                lines={lines}
                                strings={strings}
                                formFields={formFields}
                                onUpdateLine={(update,lineId) => onUpdateLine(update,lineId,groupId)}
                                onRemoveLine={(lineId) => onRemoveLine(lineId,groupId)}
                            />
                        </Fragment>
                    )
                })

                }
                <button
                    style={{margin: "12px 0 12px"}}
                    type="button"
                    className="block-input button"
                    data-confirm={strings['confirm-remove']}
                    onClick={e => {
                        e.preventDefault();
                        onRemoveConditional(id)
                    }}
                >
                    {strings['remove-condition']}
                </button>

                <AppliesToFields
                    formFields={formFields}
                    fieldsUsed={fieldsUsed}
                    fieldsNotAllowed={fieldsNotAllowed}
                    strings={strings}
                    onChange={() => {
                    }}
                    groupId={id}
                />
            </div>

        </div>
    )
};


export default Conditional;