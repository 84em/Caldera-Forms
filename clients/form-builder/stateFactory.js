import cfEditorState from '@calderajs/cf-editor-state';

/**
 * Prepare tags objects, as represented in system_values global
 *
 * @since 1.8.10
 *
 * @param tags
 * @returns {[]}
 */
function prepareTags(tags) {
    const initialValues = [];
     Object.keys(tags).forEach(tagType => {
        const tagsOfType = tags[tagType];

        if( tagsOfType.length ){
            tagsOfType.forEach(tag => {
                initialValues.push({type: tagType, tag})
            });
        }

    });
    return initialValues;
}

/**
 * Prepare field magic tags, from object supplied in global current_form_fields
 *
 * @since 1.8.10
 *
 * @param currentFormFields
 * @returns {Array}
 */
function prepareFields(currentFormFields){
    return currentFormFields && Object.keys(currentFormFields).length ? Object.keys(currentFormFields).map(
        id => {
            const field = currentFormFields[id];
            return {...field,id,tag: '%' + field.slug + '%'}
        }
    ) : [];
}
/**
 * Utility function for adding magic types from default system_tags array.
 *
 * @since 1.8.10
 *
 * @param system_values
 * @param state
 * @returns {[]}
 */
function addInitialMagicTypes(system_values,state) {
    const other = [];
    Object.keys(system_values).forEach(magicType => {
        if(undefined !== magicType && ! ['field', 'system'].includes(magicType)){
            other.push( system_values[magicType] );
        }
    });
    if( other && other.length ){
        other.forEach(o => state.addMagicType(
            o.type, true, prepareTags(o.tags)
        ));
    }
    return other;

}

/**
 * Factory for editor state management
 *
 * @since 1.8.10
 *
 * @param system_values
 * @param current_form_fields
 * @returns {{prepareSystemTags: (function(): []), createState: (function(): {getAllMagicTags: () => magicTags; getMagicTagsByType: (type: string) => magicTags; addMagicType: (typeName: string, brackets: boolean, intialTags?: (typeAndTags | undefined)) => void; removeMagicType: (typeName: string) => boolean; getAllSystemValues(): systemValues; feildEvents: EventEmitter<{addField: (fieldId: string) => void; updateField: (args: {fieldId: string; beforeUpdate: field}) => void; removeField: (field: field) => void}>; addField(field: field): boolean; getField(fieldId: string): (field | undefined); updateField(field: field): boolean; getAllFields(): fields; removeField(fieldId: string): boolean; addConditional(conditional: conditional): boolean; removeConditional(conditionalId: string): boolean; updateConditional(conditional: conditional): boolean; getConditional(conditionalId: string): (conditional | undefined); getAllConditionals(): conditionals})}}
 */
export default function (system_values,current_form_fields) {
    const systemTags = system_values && system_values.system.tags ? system_values.system.tags : [];
    //spelling error is intentional, must fix upstream.
    const intialFields = prepareFields(current_form_fields);
    const api =  {
        prepareFields,
        createState: () => {
            const state = cfEditorState({
                initialSystemValues: prepareTags(systemTags),
                intialFields
            });
            addInitialMagicTypes(system_values,state);
            return  state;

        },
        prepareSystemTags: () => prepareTags(systemTags)
    };

    return api;
}