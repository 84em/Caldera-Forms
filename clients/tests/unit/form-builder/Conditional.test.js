import React from 'react';
import Conditional, {AppliesToFields} from "../../../form-builder/components/Conditional";
import testForm from './test-form';
import renderer from 'react-test-renderer';
import system_values from "./system_values";
import stateFactory, {
    getAllFieldsUsed,
    getFieldsNotAllowedForConditional,
    getFieldsUsedByConditional
} from "../../../form-builder/stateFactory";
import {mount} from "enzyme/build";
import EnzymeAdapter from '../createEnzymeAdapter'

describe('Editor for a single conditional', () => {

    const formFields = Object.values(testForm.fields);
    const current_form_fields = formFields.map(field => {
            return {
                slug: field.slug,
                label: field.label,
                type: field.type
            }
        }
    );
    const groupId = 'con_8761120514939434';
    const factory = stateFactory(system_values, current_form_fields);
    const state = factory.createState();
    Object.values(testForm.conditional_groups.conditions).map(group => state.addConditional(factory.conditionalFromCfConfig(group)));

    const fieldsUsed = getAllFieldsUsed(state);
    const notAllowedFields = getFieldsNotAllowedForConditional(groupId, state);
    const appliedFields = getFieldsUsedByConditional(groupId, state);
    console.log(fieldsUsed.length, notAllowedFields.length, formFields.length);
    console.log(fieldsUsed);

    const strings = {'applied-fields': 'applied-fields', 'select-apply-fields': 'select-apply-fields'};
    const props = {
        onChange: jest.fn(),
        formFields,
        fieldsUsed, appliedFields, notAllowedFields, strings, groupId
    };

    test('AppliesToFields matches snapshot', () => {
        const component = renderer.create(
            <AppliesToFields
                {...props}
            />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
    test('AppliesToFields has the right number of boxes disabled', () => {
        const component = mount(
            <AppliesToFields
                {...props}
            />
        );
        expect(component.find({type: 'checkbox'}).length).toBe(6);
        expect(component.find({type: 'checkbox', disabled: true}).length).toBe(notAllowedFields.length);
    });

    test('AppliesToFields has the right number of boxes checked', () => {
        const component = mount(
            <AppliesToFields
                {...props}
            />
        );
        expect(component.find({type: 'checkbox'}).length).toBe(6);
        expect(component.find({type: 'checkbox', checked: true}).length).toBe(4);
    });


})
;
