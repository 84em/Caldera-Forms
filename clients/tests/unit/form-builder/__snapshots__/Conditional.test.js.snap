// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor for a single conditional AppliesToFields matches snapshot 1`] = `
<div
  style={
    Object {
      "float": "left",
      "paddingLeft": "12px",
      "width": "288px",
    }
  }
>
  <h4
    style={
      Object {
        "borderBottom": "1px solid rgb(191, 191, 191)",
        "margin": "0px 0px 6px; padding: 0px 0px 6px",
      }
    }
  >
    applied-fields
  </h4>
  <p
    className="description"
  >
    select-apply-fields
  </p>
  <label
    style={
      Object {
        "display": "block",
        "marginLeft": "20px",
      }
    }
  >
    Number
    <input
      checked={true}
      disabled={true}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "-20px",
        }
      }
      type="checkbox"
      value="con_8761120514939434"
    />
  </label>
  <label
    style={
      Object {
        "display": "block",
        "marginLeft": "20px",
      }
    }
  >
    Dropdown
    <input
      checked={true}
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "-20px",
        }
      }
      type="checkbox"
      value="con_8761120514939434"
    />
  </label>
  <label
    style={
      Object {
        "display": "block",
        "marginLeft": "20px",
      }
    }
  >
    Number 2
    <input
      checked={true}
      disabled={true}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "-20px",
        }
      }
      type="checkbox"
      value="con_8761120514939434"
    />
  </label>
  <label
    style={
      Object {
        "display": "block",
        "marginLeft": "20px",
      }
    }
  >
    Dropdown 2
    <input
      checked={true}
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "-20px",
        }
      }
      type="checkbox"
      value="con_8761120514939434"
    />
  </label>
  <label
    style={
      Object {
        "display": "block",
        "marginLeft": "20px",
      }
    }
  >
    Next Page
    <input
      checked={false}
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "-20px",
        }
      }
      type="checkbox"
      value="con_8761120514939434"
    />
  </label>
  <label
    style={
      Object {
        "display": "block",
        "marginLeft": "20px",
      }
    }
  >
    Back
    <input
      checked={false}
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "marginLeft": "-20px",
        }
      }
      type="checkbox"
      value="con_8761120514939434"
    />
  </label>
</div>
`;
